pkg/api/validation/validation.go

```diff
@@ -3288,6 +3290,80 @@ func ValidateNamespaceFinalizeUpdate(newNamespace, oldNamespace *api.Namespace)
 	return allErrs
 }
 
+// Set all NodeNames in the subset EndpointAddress slices to nil
+func clearNodeNamesinSubset(subset *api.EndpointSubset) {
+	for n := range subset.Addresses {
+		subset.Addresses[n].NodeName = nil
+	}
+	for n := range subset.NotReadyAddresses {
+		subset.Addresses[n].NodeName = nil
+	}
+}
+
+// ClearReadOnlyEndpointFieldsOnCreate Clears fields that are non user configurable upon create
+func ClearReadOnlyEndpointFieldsOnCreate(endpoints *api.Endpoints) {
+	for i := range endpoints.Subsets {
+		clearNodeNamesinSubset(&endpoints.Subsets[i])
+	}
+}
+
+func buildEndpointAddressToNodeNameMap(ipToNodeName map[string]string, addresses []api.EndpointAddress) {
+	// Construct lookup map of old subset IPs to NodeNames.
+	for n := range addresses {
+		_, ok := ipToNodeName[addresses[n].IP]
+		if !ok {
+			// New/unknown EndPointAddresses cannot have NodeName set but we only allow transition
+			// from nil to NodeName for pre-existing EndpointAddresses. Use empty string values to
+			// indicate IP exists but is currently set to NodeName = nil.
+			value := isExistingEndpointWithNilNodeName
+			if addresses[n].NodeName != nil {
+				value = *addresses[n].NodeName
+			}
+			ipToNodeName[addresses[n].IP] = value
+		}
+	}
+}
+
+func enforceNodeNameUnchangedEndpointAddresses(newAddresses []api.EndpointAddress, ipToNodeName map[string]string) {
+	for n := range newAddresses {
+		existingNodeName, ok := ipToNodeName[newAddresses[n].IP]
+		if !ok {
+			// New EndpointAddress not found in old - treat like Create and force it to nil
```

Comment:
We should not mutate user requests in validation.  It is either accepted or not.  In this case you have to allow updates , and moves from "" -> non-empty are valid